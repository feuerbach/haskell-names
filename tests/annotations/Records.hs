{-# LANGUAGE NamedFieldPuns, RecordWildCards #-}
module Records where

import qualified Prelude

data A = A { b :: A, c :: A } | X { b :: A, xxx :: A }

undefined = undefined

-- simple record pattern match
f A { b = q } = q

-- simple record update

fu x = x { b = b x }

-- pattern pun
g A {b} = b

-- top-level pattern pun clashes with selector
-- A { c } = undefined
ty = c

-- expression pun
tz = let b = 1 in A { b }

-- mixed pattern
h (A {b, c = (4, y)}) = (b, y)

-- mixed expression
tm = let b = 1; v = 2 in A { b, c = v }

-- stripping module name
-- (not supported by HSE yet)
{- f A { M.b } = b -}

-- simple wildcard pattern
f A { .. } = (b, c)

-- top-level wildcard pattern
Prelude.DataTypeWithSelectors { .. } = undefined
aux = (selector1,selector2)

-- mixed wildcard and normal pattern
f2 A { b = v, .. } = (v, c)

-- mixed wildcard and pun
f3 A { b, .. } = (b, c)

-- simple wildcard construction
r1 = let b = 1 in A { .. }

-- wildcard construction referring to a global and a local values
-- (global values are defined above with a top-level pattern wildcard)
r2 = let selector2 = undefined in Prelude.DataTypeWithSelectors { .. }

-- mixed simple construction and wildcard
r3 = let b = undefined in A { c = undefined, .. }

-- mixed simple construction and pun
r4 = let (b,c) = undefined in A { c, .. }

-- test that we don't bring fields from other constructors
r5 = let (b, xxx) = undefined in A { .. }
