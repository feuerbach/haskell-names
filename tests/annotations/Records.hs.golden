NamedFieldPuns at 1:14 is none
RecordWildCards at 1:30 is none
A        at  6:6 is a type or class defined here
A        at 6:10 is a value bound here
b        at 6:14 is a value bound here
A        at 6:19 is a global data type, Records.A
A        at 6:19 is a global data type, Records.A
c        at 6:22 is a value bound here
A        at 6:27 is a global data type, Records.A
A        at 6:27 is a global data type, Records.A
X        at 6:33 is a value bound here
b        at 6:37 is a value bound here
A        at 6:42 is a global data type, Records.A
A        at 6:42 is a global data type, Records.A
xxx      at 6:45 is a value bound here
A        at 6:52 is a global data type, Records.A
A        at 6:52 is a global data type, Records.A
undefined at  8:1 is a value bound here
undefined at 8:13 is a global value, Records.undefined
undefined at 8:13 is a global value, Records.undefined
f        at 11:1 is a value bound here
A        at 11:3 is a global constructor, Records.A
A        at 11:3 is a global constructor, Records.A
b        at 11:7 is a global selector, Records.b
b        at 11:7 is a global selector, Records.b
q        at 11:11 is a value bound here
q        at 11:17 is a local value defined at 11:11
q        at 11:17 is a local value defined at 11:11
fu       at 15:1 is a value bound here
x        at 15:4 is a value bound here
x        at 15:8 is a local value defined at 15:4
x        at 15:8 is a local value defined at 15:4
b        at 15:12 is a global selector, Records.b
b        at 15:12 is a global selector, Records.b
b        at 15:16 is a global selector, Records.b
b        at 15:16 is a global selector, Records.b
x        at 15:18 is a local value defined at 15:4
x        at 15:18 is a local value defined at 15:4
g        at 18:1 is a value bound here
A        at 18:3 is a global constructor, Records.A
A        at 18:3 is a global constructor, Records.A
b        at 18:6 is a global selector, Records.b
b        at 18:6 is a global selector, Records.b
b        at 18:11 is a local value defined at 18:6
b        at 18:11 is a local value defined at 18:6
ty       at 22:1 is a value bound here
c        at 22:6 is a global selector, Records.c
c        at 22:6 is a global selector, Records.c
tz       at 25:1 is a value bound here
b        at 25:10 is a value bound here
A        at 25:19 is a global constructor, Records.A
A        at 25:19 is a global constructor, Records.A
b        at 25:23 is a global selector, Records.b
b        at 25:23 is a global selector, Records.b
h        at 28:1 is a value bound here
A        at 28:4 is a global constructor, Records.A
A        at 28:4 is a global constructor, Records.A
b        at 28:7 is a global selector, Records.b
b        at 28:7 is a global selector, Records.b
c        at 28:10 is a global selector, Records.c
c        at 28:10 is a global selector, Records.c
y        at 28:18 is a value bound here
b        at 28:26 is a local value defined at 28:7
b        at 28:26 is a local value defined at 28:7
y        at 28:29 is a local value defined at 28:18
y        at 28:29 is a local value defined at 28:18
tm       at 31:1 is a value bound here
b        at 31:10 is a value bound here
v        at 31:17 is a value bound here
A        at 31:26 is a global constructor, Records.A
A        at 31:26 is a global constructor, Records.A
b        at 31:30 is a global selector, Records.b
b        at 31:30 is a global selector, Records.b
c        at 31:33 is a global selector, Records.c
c        at 31:33 is a global selector, Records.c
v        at 31:37 is a local value defined at 31:17
v        at 31:37 is a local value defined at 31:17
f        at 38:1 is a value bound here
A        at 38:3 is a global constructor, Records.A
A        at 38:3 is a global constructor, Records.A
..       at 38:7 is a record pattern wildcard which brings the following fields: Records.b, Records.c
b        at 38:15 is a local value defined at 38:7
b        at 38:15 is a local value defined at 38:7
c        at 38:18 is a local value defined at 38:7
c        at 38:18 is a local value defined at 38:7
DataTypeWithSelectors at 41:1 is a global constructor, Prelude.DataTypeWithSelectors
DataTypeWithSelectors at 41:1 is a global constructor, Prelude.DataTypeWithSelectors
..       at 41:33 is a record pattern wildcard which brings the following fields: Prelude.selector1, Prelude.selector2
undefined at 41:40 is a global value, Records.undefined
undefined at 41:40 is a global value, Records.undefined
aux      at 42:1 is a value bound here
selector1 at 42:8 is a global value, Records.selector1
selector1 at 42:8 is a global value, Records.selector1
selector2 at 42:18 is a global value, Records.selector2
selector2 at 42:18 is a global value, Records.selector2
f2       at 45:1 is a value bound here
A        at 45:4 is a global constructor, Records.A
A        at 45:4 is a global constructor, Records.A
b        at 45:8 is a global selector, Records.b
b        at 45:8 is a global selector, Records.b
v        at 45:12 is a value bound here
..       at 45:15 is a record pattern wildcard which brings the following fields: Records.c
v        at 45:23 is a local value defined at 45:12
v        at 45:23 is a local value defined at 45:12
c        at 45:26 is a local value defined at 45:15
c        at 45:26 is a local value defined at 45:15
f3       at 48:1 is a value bound here
A        at 48:4 is a global constructor, Records.A
A        at 48:4 is a global constructor, Records.A
b        at 48:8 is a global selector, Records.b
b        at 48:8 is a global selector, Records.b
..       at 48:11 is a record pattern wildcard which brings the following fields: Records.c
b        at 48:19 is a local value defined at 48:8
b        at 48:19 is a local value defined at 48:8
c        at 48:22 is a local value defined at 48:11
c        at 48:22 is a local value defined at 48:11
r1       at 51:1 is a value bound here
b        at 51:10 is a value bound here
A        at 51:19 is a global constructor, Records.A
A        at 51:19 is a global constructor, Records.A
..       at 51:23 is a record construction wildcard which assigns the following fields: Records.b = (a local value defined at 51:10)
r2       at 55:1 is a value bound here
selector2 at 55:10 is a value bound here
undefined at 55:22 is a global value, Records.undefined
undefined at 55:22 is a global value, Records.undefined
DataTypeWithSelectors at 55:35 is a global constructor, Prelude.DataTypeWithSelectors
DataTypeWithSelectors at 55:35 is a global constructor, Prelude.DataTypeWithSelectors
..       at 55:67 is a record construction wildcard which assigns the following fields: Prelude.selector1 = (a global value, Records.selector1), Prelude.selector2 = (a local value defined at 55:10)
r3       at 58:1 is a value bound here
b        at 58:10 is a value bound here
undefined at 58:14 is a global value, Records.undefined
undefined at 58:14 is a global value, Records.undefined
A        at 58:27 is a global constructor, Records.A
A        at 58:27 is a global constructor, Records.A
c        at 58:31 is a global selector, Records.c
c        at 58:31 is a global selector, Records.c
undefined at 58:35 is a global value, Records.undefined
undefined at 58:35 is a global value, Records.undefined
..       at 58:46 is a record construction wildcard which assigns the following fields: Records.b = (a local value defined at 58:10)
r4       at 61:1 is a value bound here
b        at 61:11 is a value bound here
c        at 61:13 is a value bound here
undefined at 61:18 is a global value, Records.undefined
undefined at 61:18 is a global value, Records.undefined
A        at 61:31 is a global constructor, Records.A
A        at 61:31 is a global constructor, Records.A
c        at 61:35 is a global selector, Records.c
c        at 61:35 is a global selector, Records.c
..       at 61:38 is a record construction wildcard which assigns the following fields: Records.b = (a local value defined at 61:11)
r5       at 64:1 is a value bound here
b        at 64:11 is a value bound here
xxx      at 64:14 is a value bound here
undefined at 64:21 is a global value, Records.undefined
undefined at 64:21 is a global value, Records.undefined
A        at 64:34 is a global constructor, Records.A
A        at 64:34 is a global constructor, Records.A
..       at 64:38 is a record construction wildcard which assigns the following fields: Records.b = (a local value defined at 64:11)
